In [273]:
# Imports and pip installations (if needed)
import pandas as pd


# Part 1: Load the dataset

In [274]:
# Load the given datasets
ckdNum = pd.read_csv("chronic_kidney_disease_numerical.csv")
ckdCat = pd.read_csv("chronic_kidney_disease_categorical.csv")

# Print the data
#print(ckdNum)
print(ckdCat)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: This dataset is full of different patients and their medical information. It contains several different conditions/symptoms that could be used as a way to determine whether or not someone could have Chronic Kidney Disease. The features are the columns of the datasets, and can be found in both numerical (i.e. blood pressure, red blood cell count, and age) and categorical (i.e. bacterial presence, appetite, or hypertension presence). The labels in this data are whether or not the patient has ckd, as this is the conclusion we are trying to reach.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [275]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
duplicatesN = ckdNum[ckdNum.duplicated()]
print(len(duplicatesN))
ckdNum.drop_duplicates()



# Repeat the same for categorical dataset. Print the duplicate rows and drop them
duplicatesC = ckdCat[ckdCat.duplicated()]
print(len(duplicatesC))
ckdCat.drop_duplicates()



30
30


,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
396,789906,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
397,490253,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
398,492993,3.0,0.0,NaN,normal,notpresent,notpresent,yes,no,no,good,no,yes,ckd
399,273038,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: i would assume that both datasets have all unique ids because whenever they were collecting the data they most likely got a mix of numerical and catagorical information. if there are missing rows i think it would create a new row with the id and leave all entries that it didnt have blank. 

In [276]:
# Merge the two given numerical and categorical datasets based on their unique_ID.

#Print the combined dataset
ckdFull = pd.merge(ckdNum,ckdCat, on="unique_id")
print (ckdFull)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  ...   
3       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  ...   
4       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
477     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0  ...   
478     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
479     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0  ...   
480     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
481     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0  ...   

           pc         pcc          ba  htn   dm cad

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [277]:
# Calculate the percentage of rows that contain atleast one missing value
missingRows = ckdFull[ckdFull.isnull().any(axis=1)]
percent = len(missingRows)/len(ckdFull)
# Print %
print (f'Percent of rows with missing data: {round(percent*100, 2)}%')

# Drop these rows from the dataset
ckdNoMR = ckdFull[~(ckdFull.isnull().any(axis=1))]

# Print the Dataset
print(ckdNoMR)

Percent of rows with missing data: 58.92%
     unique_id   age     bp    bgr    bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0  19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       938027  43.0   60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0  ...   
3       421471  37.0   60.0  111.0  35.0  0.8  135.0  4.1  16.2  50.0  ...   
9       109053  64.0   60.0  106.0  27.0  0.7  150.0  3.3  14.4  42.0  ...   
..         ...   ...    ...    ...   ...  ...    ...  ...   ...   ...  ...   
467     486397  20.0   70.0  123.0  44.0  1.0  135.0  3.8  14.6  44.0  ...   
469     462892  36.0   80.0   85.0  16.0  1.1  142.0  4.1  15.6  44.0  ...   
478     475200  64.0  100.0  163.0  54.0  7.2  140.0  4.6   7.9  26.0  ...   
479     168242  48.0   60.0  112.0  44.0  1.2  142.0  4.9  14.5  44.0  ...   
480     474407  21.0   90.0  107.0  40.0  1.7  125.0  3.5   8.3  23.0  ...   

           pc        

## Part 3.4: Sort the dataset according to the Labels

In [278]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
ckdNoMR.sort_values("Target")
ckdNoMR.reset_index(drop=True, inplace=True)
# Print the dataset
print(ckdNoMR)

     unique_id   age     bp    bgr    bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0  19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       938027  43.0   60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0  ...   
3       421471  37.0   60.0  111.0  35.0  0.8  135.0  4.1  16.2  50.0  ...   
4       109053  64.0   60.0  106.0  27.0  0.7  150.0  3.3  14.4  42.0  ...   
..         ...   ...    ...    ...   ...  ...    ...  ...   ...   ...  ...   
193     486397  20.0   70.0  123.0  44.0  1.0  135.0  3.8  14.6  44.0  ...   
194     462892  36.0   80.0   85.0  16.0  1.1  142.0  4.1  15.6  44.0  ...   
195     475200  64.0  100.0  163.0  54.0  7.2  140.0  4.6   7.9  26.0  ...   
196     168242  48.0   60.0  112.0  44.0  1.2  142.0  4.9  14.5  44.0  ...   
197     474407  21.0   90.0  107.0  40.0  1.7  125.0  3.5   8.3  23.0  ...   

           pc         pcc          ba  htn   dm cad appet   pe 

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [279]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
ckdEnc = pd.get_dummies(ckdNoMR, categorical_columns, dtype=int)

# Print the dataset
print(ckdEnc)

     unique_id   age     bp    bgr    bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0  19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       938027  43.0   60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0  ...   
3       421471  37.0   60.0  111.0  35.0  0.8  135.0  4.1  16.2  50.0  ...   
4       109053  64.0   60.0  106.0  27.0  0.7  150.0  3.3  14.4  42.0  ...   
..         ...   ...    ...    ...   ...  ...    ...  ...   ...   ...  ...   
193     486397  20.0   70.0  123.0  44.0  1.0  135.0  3.8  14.6  44.0  ...   
194     462892  36.0   80.0   85.0  16.0  1.1  142.0  4.1  15.6  44.0  ...   
195     475200  64.0  100.0  163.0  54.0  7.2  140.0  4.6   7.9  26.0  ...   
196     168242  48.0   60.0  112.0  44.0  1.2  142.0  4.9  14.5  44.0  ...   
197     474407  21.0   90.0  107.0  40.0  1.7  125.0  3.5   8.3  23.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  an

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [280]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
for i in numerical_columns:
    ckdEnc['Zscore'] = (ckdEnc[i] - ckdEnc[i].mean()) / ckdEnc[i].std()
    ckdEnc = ckdEnc[~(abs(ckdEnc['Zscore']) > 3)]

# Print the dataset
ckdOut = ckdEnc.drop(columns=['Zscore'])
ckdOut.reset_index()
print(ckdOut)

     unique_id   age    bp    bgr    bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0  80.0   99.0  19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0  60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       938027  43.0  60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0  ...   
3       421471  37.0  60.0  111.0  35.0  0.8  135.0  4.1  16.2  50.0  ...   
4       109053  64.0  60.0  106.0  27.0  0.7  150.0  3.3  14.4  42.0  ...   
..         ...   ...   ...    ...   ...  ...    ...  ...   ...   ...  ...   
191     155872  59.0  60.0  113.0  23.0  1.1  139.0  3.5  15.3  54.0  ...   
192     530138  23.0  80.0  111.0  34.0  1.1  145.0  4.0  14.3  41.0  ...   
193     486397  20.0  70.0  123.0  44.0  1.0  135.0  3.8  14.6  44.0  ...   
194     462892  36.0  80.0   85.0  16.0  1.1  142.0  4.1  15.6  44.0  ...   
196     168242  48.0  60.0  112.0  44.0  1.2  142.0  4.9  14.5  44.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_ye

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [281]:
# Normalize the all Numerical Attributes in the dataset.
ckdNorm = round((ckdOut - ckdOut.min()) / (ckdOut.max() - ckdOut.min()),4)

# Print the dataset
print(ckdNorm)

     unique_id     age    bp     bgr      bu      sc     sod     pot    hemo  \
0       0.1067  0.3824  0.50  0.1245  0.0682  0.0159  0.9000  0.2069  0.5172   
1       0.9346  0.4559  0.00  0.1631  0.1136  0.0952  0.8000  0.7241  1.0000   
2       0.9346  0.4559  0.00  0.1631  0.1136  0.0952  0.8000  0.7241  1.0000   
3       0.3522  0.3676  0.00  0.1760  0.1894  0.0635  0.5000  0.4138  0.8161   
4       0.0000  0.7647  0.00  0.1545  0.1288  0.0476  1.0000  0.1379  0.6092   
..         ...     ...   ...     ...     ...     ...     ...     ...     ...   
191     0.0528  0.6912  0.00  0.1845  0.0985  0.1111  0.6333  0.2069  0.7126   
192     0.4747  0.1618  0.50  0.1760  0.1818  0.1111  0.8333  0.3793  0.5977   
193     0.4254  0.1176  0.25  0.2275  0.2576  0.0952  0.5000  0.3103  0.6322   
194     0.3989  0.3529  0.50  0.0644  0.0455  0.1111  0.7333  0.4138  0.7471   
196     0.0667  0.5294  0.00  0.1803  0.2576  0.1270  0.7333  0.6897  0.6207   

        pcv  ...  cad_no  cad_yes  appe

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: Someones appetite is a strange thing to try to quantify so it seems unlikely that it can help in predicting ckd in patients.

In [282]:
#Remove that column
ckdClean = ckdEnc.drop(columns=["appet_good", "appet_poor"])
# Print the dataset
print(ckdClean)

     unique_id   age    bp    bgr    bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0  80.0   99.0  19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0  60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       938027  43.0  60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0  ...   
3       421471  37.0  60.0  111.0  35.0  0.8  135.0  4.1  16.2  50.0  ...   
4       109053  64.0  60.0  106.0  27.0  0.7  150.0  3.3  14.4  42.0  ...   
..         ...   ...   ...    ...   ...  ...    ...  ...   ...   ...  ...   
191     155872  59.0  60.0  113.0  23.0  1.1  139.0  3.5  15.3  54.0  ...   
192     530138  23.0  80.0  111.0  34.0  1.1  145.0  4.0  14.3  41.0  ...   
193     486397  20.0  70.0  123.0  44.0  1.0  135.0  3.8  14.6  44.0  ...   
194     462892  36.0  80.0   85.0  16.0  1.1  142.0  4.1  15.6  44.0  ...   
196     168242  48.0  60.0  112.0  44.0  1.2  142.0  4.9  14.5  44.0  ...   

     dm_yes  cad_no  cad_yes  pe_no  pe_yes  ane_no  ane_yes  Target_ckd  \

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [283]:
# Export the dataframe to a new csv file
ckdClean.to_csv("CkdClean.csv", index=False)

# Export the dataframe to a new json file
ckdClean.to_json("CkdClean.json", index=False)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

### Paste the markdown table here

| unique_id | age | bp | bgr | bu | sc | sod | pot | hemo | pcv | wbcc | rbcc | al | su | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|-----------|-------|-------|--------|-------|-------|-------|--------|--------|--------|--------|--------|-------|-------|----------------|---------------|----------------|-------------|------------------|---------------|-----------------|--------------|----------|-----------|---------|----------|----------|-----------|---------------|---------------|---------|----------|----------|-----------|---------------|----------------|
| 0.1067 | 0.3824 | 0.5 | 0.1245 | 0.0682 | 0.0159 | 0.9 | 0.2069 | 0.5172 | 0.6154 | 0.2752 | 0.9697 | 0.0 | 0.0 | 0.0 | 1.0 | 0.0 | 1.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 |
| 0.9346 | 0.4559 | 0.0 | 0.1631 | 0.1136 | 0.0952 | 0.8 | 0.7241 | 1.0 | 0.5769 | 0.2661 | 0.697 | 0.0 | 0.0 | 0.0 | 1.0 | 0.0 | 1.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 |
| 0.9346 | 0.4559 | 0.0 | 0.1631 | 0.1136 | 0.0952 | 0.8 | 0.7241 | 1.0 | 0.5769 | 0.2661 | 0.697 | 0.0 | 0.0 | 0.0 | 1.0 | 0.0 | 1.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 |
| 0.3522 | 0.3676 | 0.0 | 0.176 | 0.1894 | 0.0635 | 0.5 | 0.4138 | 0.8161 | 0.8462 | 0.1101 | 0.7576 | 0.0 | 0.0 | 0.0 | 1.0 | 0.0 | 1.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 |
| 0.0 | 0.7647 | 0.0 | 0.1545 | 0.1288 | 0.0476 | 1.0 | 0.1379 | 0.6092 | 0.5385 | 0.3486 | 0.4545 | 0.0 | 0.0 | 0.0 | 1.0 | 0.0 | 1.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 |
| 0.5875 | 0.6765 | 0.5 | 0.2618 | 0.0606 | 0.1111 | 0.7 | 0.2069 | 0.7701 | 0.9615 | 0.2294 | 0.8788 | 0.0 | 0.0 | 0.0 | 1.0 | 0.0 | 1.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 |
| 0.7127 | 0.1618 | 0.0 | 0.1073 | 0.1061 | 0.0635 | 0.8333 | 0.7241 | 0.6782 | 0.9231 | 0.1835 | 0.4242 | 0.0 | 0.0 | 0.0 | 1.0 | 0.0 | 1.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 |
| 0.4926 | 0.5441 | 0.5 | 0.2232 | 0.1667 | 0.127 | 0.6333 | 0.3448 | 0.908 | 0.5 | 0.1193 | 0.5152 | 0.0 | 0.0 | 0.0 | 1.0 | 0.0 | 1.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 | 0.0 | 1.0 |
| 0.3962 | 0.5147 | 0.0 | 0.1674 | 0.1136 | 0.1111 | 0.7 | 0.6207 | 0.7701 | 0



This is the output i recieved from 383GPT

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Answer: It seems to struggle with the amount of information that is there so if you feed it smaller chunks of data it might work better. However this is much more time consuming, so you may just be better off writing code to parse a csv file into this format.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [284]:
# Code to rename all the columns in the dataset
# Mapping dictionary
abbreviation_mapping = {
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia'
}

# Rename the columns using the mapping
#i am using ckd full here because after encoding the names of the columns were changed
#ckd full is the dataframe after merging the two tables
ckdFull.rename(columns=abbreviation_mapping, inplace=True)

# Display the updated DataFrame with full column names
print(ckdFull)

     unique_id   age  blood pressure  blood glucose random  blood urea  \
0       203694  38.0            80.0                  99.0        19.0   
1       938027  43.0            60.0                 108.0        25.0   
2       938027  43.0            60.0                 108.0        25.0   
3       421471  37.0            60.0                 111.0        35.0   
4       764115  70.0            90.0                 144.0       125.0   
..         ...   ...             ...                   ...         ...   
477     522912  56.0            80.0                 165.0        55.0   
478     475200  64.0           100.0                 163.0        54.0   
479     168242  48.0            60.0                 112.0        44.0   
480     474407  21.0            90.0                 107.0        40.0   
481     926033  50.0            70.0                 230.0        50.0   

     serum creatinine  sodium  potassium  hemoglobin  packed cell volume  ...  \
0                 0.5   147.0 

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [285]:
# Converted SQL to Pandas code
# Sample DataFrame
data = {
    'Target': ['ckd', 'notckd', 'ckd', 'notckd', 'ckd', 'notckd'],
    'other_column': [1, 2, 3, 4, 5, 6]
}
df = pd.DataFrame(data)

# Using groupby and size to get the count of each Target
result = df.groupby('Target').size().reset_index(name='count')

# Display the result
print(result)

   Target  count
0     ckd      3
1  notckd      3


*Describe what the above code does here*

The Code above splits the dataframe into two seperate partitions sorted by either ckd or notckd and then counts the total number in each partition, and finally prints it to the console.